Added 2nd order Placzek correction to POLARIS Total Scattering

sf1919 opened the pull request:
**Description of work.**

Add support for the 2nd order term in the Placzek correction to the ISIS total scattering workflow. The second order term was added by the ORNL team in the algorithm CalculatePlaczek so this task plumbs that into the ISIS workflow.

As discussed with Helen Playford

**To test:**

Run this script and check it completes successfully. It is a varation of the script on https://github.com/mantidproject/mantid/pull/34193. There is a zip file attached to that PR containing some config files that you'll need for this to work. The `config_file_path` in the script below needs updating to match the location where you unzip the files:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

from isis_powder import polaris, SampleDetails

config_file_path = r"C:\Total Scattering\polaris_config_example.yaml"
polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF")

sample_details = SampleDetails(height=4.0, radius=0.2985, center=[0, 0, 0], shape='cylinder')
sample_details.set_material(chemical_formula='Si', packing_fraction=0.6)
polaris.set_sample_details(sample=sample_details)

polaris.create_vanadium(first_cycle_run_no="98532", multiple_scattering=False)

polaris.focus(run_number="98533", input_mode='Summed')

q_lim = [[2.5, 3, 4, 6, 7], [3.5, 5, 7, 11, 40]]

# default params for system test
polaris.create_total_scattering_pdf(run_number="98533", merge_banks=False,
                                    pdf_type="G(r)", lorch_filter=True, debug=True, 
                                    placzek_order=2, sample_temp=273)
```

Fixes #34196.


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

sf1919 committed:
Move to using CalculatePlaczek algorithm

sf1919 committed:
Enabled use of 2nd order Placzek Corrections

- includes setting default of order 1 when variable not provided or not 1 or 2
- checks that a suitable sample temperature has been provided if placzek_order=2

DannyHindson commented on scripts/Diffraction/isis_powder/polaris_routines/polaris_algs.py:
```diff
@@ -91,13 +91,17 @@ def generate_ts_pdf(run_number, focus_file_path, sample_details, merge_banks=Fal
     raw_ws = mantid.Load(Filename='POLARIS'+str(run_number))
     sample_geometry_json = sample_details.generate_sample_geometry()
+    # import pydevd_pycharm
+    # pydevd_pycharm.settrace('localhost', port=8080, stdoutToServer=True, stderrToServer=True)
```
These 2 debug lines need removing

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/TotScatCalculateSelfScattering.py:
```diff
@@ -75,10 +79,19 @@ def PyExec(self):
+                    placzek_kwargs.update({'SampleTemperature': float(sample_temp)})
+            except ValueError:
+                raise RuntimeError("To calculate the second order Placzek correction you must provide "
+                                   "a sample_temp.")
```
The CalculatePlaczek algorithm checks whether a temperature has been supplied for order=2 so I'm wondering if we need to do a validation here as well or whether we can just leave it to the CalculatePlaczek algorithm? The validation in the algorithm also checks if a temperature is present in the sample logs.

If you do leave this validation code in place perhaps this should be "...you must provide a sample temperature" rather than using the variable name.

DannyHindson reviewed (commented):
Overall looks good. I've just added a couple of minor comments

sf1919 committed:
Modifications to CalculatePlaczek

- removed line that set the Distribution back to False erroneously
- Changed validation for Sample Temperature from Sample logs to prevent crash
- Changed validation logic for checking if spectra are empty, only if the minimum number of spectra have been provided

sf1919 committed:
Changes after initial review

- removed commented out code
- enabled algorithm to defer to CalculatePlaczek for Sample Temperature if none provided by user

sf1919 committed:
Further changes to CalculatePlaczek

- simplified checks for sample temperature in logs in entire file.

sf1919 committed:
Added documentation

- Updated TotScatCalculateSelfScattering-v1 documentation and workflow diagram
- Updated Polaris.create_total_scattering_pdf documentation in Polaris techniques documentation
- Added release note

sf1919 committed:
Added tests

DannyHindson committed:
Add 2nd order derivative calc to FitIncidentSpectrum

Needed to support the 2nd order Placzek correction in the
ISIS total scattering workflow
This is approximately the same as the change done by ORNL in
their total scattering workflow
Remove code for old scipy version from FitIncidentSpectrum because
it looked wrong (missing factor of 2) and the "old" version is now
more than 5 years old so won't ever be used

DannyHindson committed:
Remove validation on temperature from workflow algorithm

Already done by CalculatePlaczek algorithm

DannyHindson committed:
Some further documentation updates relating to FitIncidentSpectrum

DannyHindson reviewed (approved):
All looks good now. Thanks for your help Sarah on this and the other issues during your time with the diffraction team!
